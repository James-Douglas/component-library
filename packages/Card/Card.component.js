import React from 'react';
import PropTypes from 'prop-types';
import { spacingPropTypes } from '@comparethemarketau/manor-utils';
import { useId } from '@comparethemarketau/manor-hooks';
import StyledCard from './Card.styles';

const Card = ({
  id: propsId,
  children,
  margin,
  padding,
  className,
}) => {
  const id = useId(propsId);
  return (
    <StyledCard id={id} margin={margin} padding={padding} className={className}>
      {children}
    </StyledCard>
  );
};

Card.propTypes = {
  /**
   * Unique identifier for the card
   */
  id: PropTypes.string,
  /**
   * Card content
   */
  children: PropTypes.node,
  /**
   * Allows setting of margin values for the card
   */
  margin: spacingPropTypes,
  /**
   * Allows setting of padding values for the card
   */
  padding: spacingPropTypes,
  /**
   * Classes to be applied to the Card component
   */
  className: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
};

Card.defaultProps = {
  id: null,
  children: [],
  margin: ['4', '4', '16'],
  padding: ['16'],
  className: '',
};

export default Card;
