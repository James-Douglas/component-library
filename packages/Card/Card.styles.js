import styled from 'styled-components';
import { applySpacing } from '@comparethemarketau/manor-utils';

const StyledCard = styled.div`
  box-sizing: border-box;
  display: flex;
  width: 100%;
  min-width: ${({ theme }) => theme.spacing['80']};
  min-height: ${({ theme }) => theme.spacing['80']};
  background: ${({ theme }) => theme.card.background};
  box-shadow: ${({ theme }) => theme.card.shadow};
  border: ${({ theme }) => theme.card.border};
  ${({ margin, theme }) => applySpacing(theme, 'margin', margin)};
  ${({ padding, theme }) => applySpacing(theme, 'padding', padding)};
`;

export default StyledCard;
