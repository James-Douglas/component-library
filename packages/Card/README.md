# Card

## Installation

`yarn add @comparethemarketau/manor-card`


## Documentation

For technical documentation, see [Storybook](https://services.dev.comparethemarket.cloud/manor/?path=/docs/components-cards--card).
Design system documentation can be found here [Bento](https://zeroheight.com/9942937b5/p/377875-cards/b/97d6bd)
