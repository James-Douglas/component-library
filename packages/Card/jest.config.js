const baseConfig = require('../../jest.config.base');

module.exports = {
  ...baseConfig,
  roots: [
    '<rootDir>/packages/Card',
  ],
  collectCoverageFrom: [
    '*.{js,jsx,ts,tsx}',
  ],
  name: 'Card',
  displayName: 'Card',
  rootDir: '../..',
};
