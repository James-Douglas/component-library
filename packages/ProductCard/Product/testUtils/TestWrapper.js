import React from 'react';
import ProductProvider from '../../Context/ProductProvider';

const ProductTestWrapper = ({
  // eslint-disable-next-line react/prop-types
  children, product, onDetails = () => {}, onSelect = () => {},
}) => (
  <ProductProvider product={product} onDetails={onDetails} onSelect={onSelect}>
    { children }
  </ProductProvider>
);

export default ProductTestWrapper;
