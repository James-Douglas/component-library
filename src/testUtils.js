// test-utils.js
import {
  render, fireEvent, wait, act,
} from '@testing-library/react';
import ManorProvider from './providers/ManorProvider';

const customRender = (ui, options) => render(ui, { wrapper: ManorProvider, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export {
  customRender as render,
  // export other used functions to appease eslint
  fireEvent,
  wait,
  act,
};
