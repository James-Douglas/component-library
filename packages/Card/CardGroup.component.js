import React from 'react';
import PropTypes from 'prop-types';
import { useId } from '@comparethemarketau/manor-hooks';
import { spacingPropTypes } from '@comparethemarketau/manor-utils';
import { ManorProvider } from '@comparethemarketau/manor-provider';
import Card from './Card.component';
import { StyledCardGroup, StyledCardGroupChildren } from './CardGroup.styles';

const getChildren = (children, cols, { margin, padding }, className) => children.map((child, index) => (
  // eslint-disable-next-line react/no-array-index-key
  <StyledCardGroupChildren cols={cols} key={`card-group-${index}`}>
    {React.cloneElement(child, { margin, padding, className })}
  </StyledCardGroupChildren>
));

const CardGroup = ({
  id: propsId,
  cols,
  children,
  cardProps,
  className,
  theme,
}) => {
  const id = useId(propsId);
  return (
    <ManorProvider theme={theme}>
      <StyledCardGroup id={id}>
        {getChildren(children, cols, cardProps, className)}
      </StyledCardGroup>
    </ManorProvider>
  );
};

CardGroup.propTypes = {
  /**
   * Unique identifier for the CardGroup
   */
  id: PropTypes.string,
  /**
   * Number of cols to spread cards into
   */
  cols: PropTypes.oneOf([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]),
  /**
   * Card Components to be contained in this CardGroup
   */
  children: PropTypes.oneOfType([
    PropTypes.shape({
      type: PropTypes.oneOf([Card]),
    }),
    PropTypes.arrayOf(
      PropTypes.shape({
        type: PropTypes.oneOf([Card]),
      }),
    ),
  ]),
  /**
   * Can be used to set margin & padding on all cards within the CardGroup.
   */
  cardProps: PropTypes.shape({
    margin: spacingPropTypes,
    padding: spacingPropTypes,
  }),
  /**
   * Classes to be applied to the CardGroup component
   */
  className: PropTypes.string,
  // eslint-disable-next-line react/forbid-prop-types
  /**
   * Manor theme, if not provided the ctm theme will be used.
   */
  // eslint-disable-next-line react/forbid-prop-types
  theme: PropTypes.object,
};

CardGroup.defaultProps = {
  id: null,
  cols: 1,
  children: [],
  cardProps: {},
  className: '',
  theme: undefined,
};

export default CardGroup;
