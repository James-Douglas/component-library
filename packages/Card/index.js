import Card from './Card.component';
import CardGroup from './CardGroup.component';

export { Card, CardGroup };
