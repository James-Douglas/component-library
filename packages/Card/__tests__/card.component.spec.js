import React from 'react';
import { render } from '../../../testUtils';
import Card from '../Card.component';

describe('Card', () => {
  it('renders', () => {
    const { container } = render(<Card id="card-one" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders with margin', () => {
    const { container } = render(<Card id="card-one" margin={['4', '8', '12', '16']} />);

    const card = container.querySelector('[id="card-one"]');
    expect(card).toHaveStyle('margin: 0.4rem 0.8rem 1.2rem 1.6rem');
  });

  it('renders with padding', () => {
    const { container } = render(<Card id="card-one" padding={['4', '8', '12', '16']} />);

    const card = container.querySelector('[id="card-one"]');
    expect(card).toHaveStyle('padding: 0.4rem 0.8rem 1.2rem 1.6rem');
  });

  it('renders with margin and padding', () => {
    const { container } = render(<Card id="card-one" margin={['4', '8', '12', '16']} padding={['8', '8', '8', '4']} />);

    const card = container.querySelector('[id="card-one"]');
    expect(card).toHaveStyle('margin: 0.4rem 0.8rem 1.2rem 1.6rem');
    expect(card).toHaveStyle('padding: 0.8rem 0.8rem 0.8rem 0.4rem');
  });

  it('renders with box-shadow', () => {
    const { container } = render(<Card id="card-one" boxShadow={false} />);

    const card = container.querySelector('[id="card-one"]');
    expect(card).toHaveStyle('box-shadow: none');
  });

  it('renders without box-shadow', () => {
    const { container } = render(<Card id="card-one" />);

    const card = container.querySelector('[id="card-one"]');
    expect(card).not.toHaveStyle('box-shadow: none');
  });
});
