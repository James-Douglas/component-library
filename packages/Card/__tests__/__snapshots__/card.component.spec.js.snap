// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders 1`] = `"<div id=\\"card-one\\" class=\\"sc-AxjAm dqYQfW\\"></div>"`;
